# Shootings in America - Data Cleaning

In this project I take a closer look at shootings in the United States of America. In the recent past, gun violence in America has been making news headlines. This is concerning even for someone living thousands of miles away. I found a dataset on Kaggle containing data on shootings from 1982 to the current year 2023 and decided to jump right in.

You are welcome to tag along.

In [1]:
# importing the usual suspects

import numpy as np
# linear algebra, scientific computing

import pandas as pd
# data processing, manipulation and analysis

import matplotlib.pyplot as plt
# data visualization

import seaborn as sns
# data visualization

In [2]:
# reading in the data

df=pd.read_csv('shooting-1982-2023.csv')

In [3]:
df.head()

# checking the first 5 rows of the data

case                   location      date  \
0  Nashville religious school shooting              Nashville, TN   3-27-23   
1   Michigan State University shooting     East Lansing, Michigan   2-13-23   
2         Half Moon Bay spree shooting  Half Moon Bay, California   1-23-23   
3        LA dance studio mass shooting  Monterey Park, California   1-21-23   
4            Virginia Walmart shooting       Chesapeake, Virginia  11-22-22   

                                             summary  fatalities injured  \
0  Audrey Hale, 28, who was a former student at t...           6       1   
1  Anthony D. McRae, 43, opened fire at Berkey Ha...           3       5   
2  Chunli Zhao, 67, suspected of carrying out the...           7       1   
3  Huu Can Tran, 72, fled the scene in a white va...          11      10   
4  Andre Bing, 31, who worked as a supervisor at ...           6       6   

  total_victims location.1 age_of_shooter prior_signs_mental_health_issues  \
0             6     School             28                                -   
1             8     School             43                                -   
2             8  workplace             67                                -   
3            21      Other             72                              yes   
4            12  Workplace             31                                -   

   ... weapons_obtained_legally where_obtained  \
0  ...                        -              -   
1  ...                      yes              -   
2  ...                        -              -   
3  ...                        -              -   
4  ...                        -              -   

                                      weapon_type weapon_details   race  \
0      semiautomatic rifle, semiautomatic handgun              -      -   
1                          semiautomatic handguns              -  Black   
2                           semiautomatic handgun              -  Asian   
3  semiautomatic assault weapon (Details pending)              -  Asian   
4                           semiautomatic handgun              -  Black   

                                      gender latitude longitude   type  year  
0  F (identified as transgender, per police)        -         -   Mass  2023  
1                                          M        -         -   Mass  2023  
2                                          M        -         -  Spree  2023  
3                                          M        -         -   Mass  2023  
4                                          M        -         -   Mass  2022  

[5 rows x 21 columns]

In [4]:
df.info()

# checking the columns and their data types as well as whether we have null values.

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 141 entries, 0 to 140
Data columns (total 21 columns):
 #   Column                            Non-Null Count  Dtype 
---  ------                            --------------  ----- 
 0   case                              141 non-null    object
 1   location                          141 non-null    object
 2   date                              141 non-null    object
 3   summary                           141 non-null    object
 4   fatalities                        141 non-null    int64 
 5   injured                           141 non-null    object
 6   total_victims                     141 non-null    object
 7   location.1                        141 non-null    object
 8   age_of_shooter                    141 non-null    object
 9   prior_signs_mental_health_issues  141 non-null    object
 10  mental_health_details             141 non-null    object
 11  weapons_obtained_legally          141 non-null    object
 12  where_obtained        

We have 21 columns and 141 observations.

The column 'weapon_details' is missing a single observation. We'll figure out how to deal with that later.

A few columns are not in the data type we need for the purposes of this project so we have to take care of that.

In [5]:
df.dtypes

case                                object
location                            object
date                                object
summary                             object
fatalities                           int64
injured                             object
total_victims                       object
location.1                          object
age_of_shooter                      object
prior_signs_mental_health_issues    object
mental_health_details               object
weapons_obtained_legally            object
where_obtained                      object
weapon_type                         object
weapon_details                      object
race                                object
gender                              object
latitude                            object
longitude                           object
type                                object
year                                 int64
dtype: object

In [6]:
df.columns

Index(['case', 'location', 'date', 'summary', 'fatalities', 'injured',
       'total_victims', 'location.1', 'age_of_shooter',
       'prior_signs_mental_health_issues', 'mental_health_details',
       'weapons_obtained_legally', 'where_obtained', 'weapon_type',
       'weapon_details', 'race', 'gender', 'latitude', 'longitude', 'type',
       'year'],
      dtype='object')

## Cleaning the data

In [7]:
df['case'].head()

0    Nashville religious school shooting
1     Michigan State University shooting
2           Half Moon Bay spree shooting
3          LA dance studio mass shooting
4              Virginia Walmart shooting
Name: case, dtype: object

A bunch of strings and zero nulls so we leave this column as is.

In [8]:
df['location'].head()

0                Nashville, TN
1       East Lansing, Michigan
2    Half Moon Bay, California
3    Monterey Park, California
4         Chesapeake, Virginia
Name: location, dtype: object

Contains both the city and state

In [9]:
# let's split the 'location' column into 2 columns - 'city' and 'state'

df['city']=df['location'].apply(lambda x: x.split(',')[0])
df['state']=df['location'].apply(lambda x: x.split(',')[1])

In [10]:
df[['city','state']].head()

# confirming that what we did above worked

city        state
0      Nashville           TN
1   East Lansing     Michigan
2  Half Moon Bay   California
3  Monterey Park   California
4     Chesapeake     Virginia

In [11]:
# a closer look at the new 'state' column

df['state'].unique()

array([' TN', ' Michigan', ' California', ' Virginia', ' Colorado',
       ' North Carolina', ' Indiana', ' Illinois', ' Alabama',
       ' Maryland', ' Oklahoma', ' Texas', ' New York', ' Georgia',
       ' Missouri', ' Wisconsin', ' New Jersey', ' Florida', ' Ohio',
       ' Pennsylvania', ' Tennessee', ' Nevada', ' Washington',
       ' Lousiana', ' Kansas', ' Oregon', ' South Carolina', ' D.C.',
       ' Connecticut', ' Minnesota', ' Arizona', ' Kentucky', ' Nebraska',
       ' Utah', ' Mississippi', ' Massachusetts', ' Hawaii', ' Arkansas',
       ' Iowa'], dtype=object)

In [12]:
# TN and D.C. are abbreviations while the rest are entered in full
# for the sake of consistency we will write them in full below

df['state']=df['state'].apply(lambda x: x.replace('TN','Tennessee').replace('D.C','District of Columbia'))

In [13]:
df['date'].head()

0     3-27-23
1     2-13-23
2     1-23-23
3     1-21-23
4    11-22-22
Name: date, dtype: object

In [14]:
# as seen above, 'date' column is a string so we convert it to datetime

df['date']=pd.to_datetime(df['date'])

In [15]:
# now that 'date' is in the appropriate type, we split it into two - 'month' and 'day' - columns

df['month']=df['date'].dt.month
df['day']=df['date'].dt.day

In [16]:
df[['month','day']].tail()

# confirming that what we did above worked

month  day
136      4   23
137      8   20
138      7   18
139      6   29
140      8   20

In [17]:
# the months are numeric so we map them to their corresponding names

df['month']=df['month'].map({1:'January',2:'February',3:'March',4:'April',5:'May',6:'June',7:'July',8:'August',9:'September',10:'October',11:'November',12:'December',})

In [18]:
df['summary'].head()

0    Audrey Hale, 28, who was a former student at t...
1    Anthony D. McRae, 43, opened fire at Berkey Ha...
2    Chunli Zhao, 67, suspected of carrying out the...
3    Huu Can Tran, 72, fled the scene in a white va...
4    Andre Bing, 31, who worked as a supervisor at ...
Name: summary, dtype: object

A bunch of strings and zero nulls so we leave this column as is.

In [19]:
df['fatalities'].unique()

array([ 6,  3,  7, 11,  5,  4, 21, 10,  9,  8, 22, 12, 17, 26, 58, 49, 14,
       27, 13, 32, 24, 15], dtype=int64)

In [20]:
df['injured'].unique()

array(['1', '5', '10', '6', '25', '2', '46', '0', '(TK - "fewer than 10"',
       '17', '3', '7', '8', '27', '26', '12', '4', '22', '13', '20',
       '546', '11', '53', '14', '21', '9', '70', '31', '23', '24', '19',
       '29'], dtype=object)

This one needs cleaning. One of the values is strange and the data type is incorrect.

In [21]:
df[df['injured']=='(TK - "fewer than 10"']['summary']

12    Michael Louis, 45, killed four, including two ...
Name: summary, dtype: object

A Google search of the incident and the incident summary above don't shed light on the number of injured people.

I've decided to drop the entire row.

In [22]:
df=df.drop(12,axis=0)

In [23]:
df['injured'].unique()

array(['1', '5', '10', '6', '25', '2', '46', '0', '17', '3', '7', '8',
       '27', '26', '12', '4', '22', '13', '20', '546', '11', '53', '14',
       '21', '9', '70', '31', '23', '24', '19', '29'], dtype=object)

In [24]:
df['total_victims'].unique()

array(['6', '8', '21', '12', '30', '5', '7', '53', '3', '4', '38', '13',
       '11', '9', '15', '10', '32', '36', '48', '16', '34', '17', '23',
       '46', '604', '102', '35', '18', '19', '20', '29', '82', '44', '26',
       '55', '22', '37', '28', '25', '14', '41'], dtype=object)

In [25]:
# changing 'injured' and 'total_victims' from strings to integers

df['injured']=df['injured'].astype('int64')
df['total_victims']=df['total_victims'].astype('int64')

In [26]:
df[['injured','total_victims']].dtypes

injured          int64
total_victims    int64
dtype: object

Perfect!

In [27]:
df['location.1'].unique()

array(['School', 'workplace', 'Other', 'Workplace', 'religious',
       'Religious', 'Military', 'Airport', '\nWorkplace', 'Other\n'],
      dtype=object)

This needs a bit of work.

In [28]:
# cleaning

df['location.1']=df['location.1'].apply(lambda x: x.title().replace('\n',''))

In [29]:
df['location.1'].unique()

array(['School', 'Workplace', 'Other', 'Religious', 'Military', 'Airport'],
      dtype=object)

Done.

In [30]:
df['age_of_shooter'].unique()

array(['28', '43', '67', '72', '31', '22', '15', '20', '21', '70', '23',
       '18', '-', '57', '19', '44', '51', '36', '24', '40', '45', '32',
       '46', '26', '54', '29', '38', '17', '47', '37', '64', '39', '25',
       '33', '27', '34', '42', '59', '41', '52', '16', '48', '66', '11',
       '35', '55', '50'], dtype=object)

In [31]:
# trying to glean the age of the shooter from the incident summary

df[df['age_of_shooter']=='-']['summary']

15    "A man believed to be meeting his three childr...
24    David N. Anderson, 47, and Francine Graham, 50...
Name: summary, dtype: object

In [32]:
# 15 - a Google search revealed this shooter was 39

df.at[15,'age_of_shooter']='39' 

# 24 - from the summary, we have two shooters and their ages but I've decide to go with the one who appears first

df.at[24,'age_of_shooter']='47' 

In [33]:
# changing the type to int

df['age_of_shooter']=df['age_of_shooter'].astype('int64')

In [34]:
df['age_of_shooter'].unique()

array([28, 43, 67, 72, 31, 22, 15, 20, 21, 70, 23, 18, 39, 57, 19, 44, 51,
       47, 36, 24, 40, 45, 32, 46, 26, 54, 29, 38, 17, 37, 64, 25, 33, 27,
       34, 42, 59, 41, 52, 16, 48, 66, 11, 35, 55, 50], dtype=int64)

In [35]:
# we can create age groups from the age column

def age_group(age):
    if age<13:
        return 'Child'
    elif age<20:
        return 'Teenager'
    elif age<35:
        return 'Young Adult'
    elif age<60:
        return 'Middle Aged'
    else:
        return 'Senior'

In [36]:
df['age_group']=df['age_of_shooter'].apply(age_group)

df['age_group'].unique()

array(['Young Adult', 'Middle Aged', 'Senior', 'Teenager', 'Child'],
      dtype=object)

Nicely done.

In [37]:
df['prior_signs_mental_health_issues'].unique()

array(['-', 'yes', 'TBD', 'Yes', 'Unclear', 'Unknown', 'No', 'Unclear '],
      dtype=object)

In [38]:
# fixing the inconsistencies

df['prior_signs_mental_health_issues']=df['prior_signs_mental_health_issues'].apply(
    lambda x: x.replace('-','Unknown').
    replace('yes','Yes').replace(' ','').
    replace('Unclear','Unknown').
    replace('TBD','Unknown'))

df['prior_signs_mental_health_issues'].unique()

array(['Unknown', 'Yes', 'No'], dtype=object)

In [39]:
df['mental_health_details'].head()

0                                                    -
1                                                    -
2                                                    -
3    According to the LA Times, "Two law enforcemen...
4                                                    -
Name: mental_health_details, dtype: object

Let's ignore the '-' since this columns just contains a bunch of descriptions.

In [40]:
df['weapons_obtained_legally'].unique()

array(['-', 'yes', 'Yes', 'No', 'TBD',
       'Kelley passed federal criminal background checks; the US Air Force failed to provide information on his criminal history to the FBI',
       'Unknown', '\nYes',
       'Yes ("some of the weapons were purchased legally and some of them may not have been")',
       'Yes '], dtype=object)

Some janitorial work is needed here.

In [41]:
df['weapons_obtained_legally']=df['weapons_obtained_legally'].apply(
    lambda x: x.replace('-','Unknown').replace('yes','Yes').replace('\n','').replace('Yes ','Yes').replace('TBD','Unknown').
    replace('Kelley passed federal criminal background checks; the US Air Force failed to provide information on his criminal history to the FBI','TBD').replace('Yes("some of the weapons were purchased legally and some of them may not have been")','Yes'))

df['weapons_obtained_legally'].unique()

array(['Unknown', 'Yes', 'No', 'TBD'], dtype=object)

In [42]:
df['weapons_obtained_legally']=df['weapons_obtained_legally'].apply(lambda x: x.replace('TBD','Unknown'))
    
df['weapons_obtained_legally'].unique()

array(['Unknown', 'Yes', 'No'], dtype=object)

Done!

In [43]:
df['where_obtained'].head()

0    -
1    -
2    -
3    -
4    -
Name: where_obtained, dtype: object

Let's ignore the hyphens since this columns just contains a bunch of descriptions.

Moving along swiftly...

In [44]:
df['weapon_type'].head()

0        semiautomatic rifle, semiautomatic handgun
1                            semiautomatic handguns
2                             semiautomatic handgun
3    semiautomatic assault weapon (Details pending)
4                             semiautomatic handgun
Name: weapon_type, dtype: object

In [45]:
df['weapon_details'].sample(frac=0.1)

129      .22-caliber Ruger sawed-off semiautomatic rifle
93     Winchester 1300 pump-action shotgun; .357 Magn...
79     10mm Glock, 9mm SIG Sauer P226 semiautomatic h...
60     Sig Sauer MCX rifle, Glock 17 9mm; high-capaci...
104    .40-caliber Glock 23, .22-caliber Ruger semiau...
8                                   Sig Sauer M400 rifle
48                                                     -
85                     .45-caliber semiautomatic handgun
55                                         .357 revolver
40                          12-gauge pump-action shotgun
133    Two Intratec MAC-11, 9mm SIG Sauer semiautomat...
54                                                     -
99     9mm Glock 19, .22-caliber Walther P22 semiauto...
62                      9 mm handgun (ammo used unclear)
Name: weapon_details, dtype: object

In [46]:
# remember we had an empty observation in this column

df[df['weapon_details'].isnull()]

case                   location       date  \
7  Raleigh spree shooting  Hedingham, North Carolina 2022-10-13   

                                             summary  fatalities  injured  \
7  Austin Thompson, 15, went on a rampage in the ...           5        2   

   total_victims location.1  age_of_shooter prior_signs_mental_health_issues  \
7              7      Other              15                          Unknown   

   ... gender latitude longitude   type  year       city            state  \
7  ...      M        -         -  Spree  2022  Hedingham   North Carolina   

     month day age_group  
7  October  13  Teenager  

[1 rows x 26 columns]

A Google search of this incident doesn't help so let's leave it as is.

In [47]:
df['race'].unique()

array(['-', 'Black', 'Asian', 'White', 'Latino', 'Other', 'White ',
       'Native American', 'white', 'black', 'unclear'], dtype=object)

In [48]:
df[df['race']=='-']['summary']

0     Audrey Hale, 28, who was a former student at t...
11    Joe Louis Esquivel, 23, shot four coworkers an...
15    "A man believed to be meeting his three childr...
16    Ethan Crumbley, a 15-year-old student at Oxfor...
17    Samuel Cassidy, 57, a Valley Transportation Au...
19    Aminadab Gaxiola Gonzalez, 44, allegedly opene...
20    Ahmad Al Aliwi Alissa, 21, carried out a mass ...
22    Joaquin S. Roman, 31, went on a rampage culmin...
25    Ahmed Mohammed al-Shamrani, 21, a Saudi Arabia...
29    Santino William LeGan, 19, fired indiscriminat...
35    Ian David Long, 28, dressed in black and armed...
53    John Robert Neumann, Jr., 45, a former employe...
57    Arcan Cetin, 20, killed a teen girl and three ...
Name: summary, dtype: object

We could attempt to infer race from the name in the incident summary but I don't think we should.

In [49]:
df['race']=df['race'].apply(lambda x: x.title().replace('-','Unclear').replace('White ','White').replace('Other','Unclear'))

df['race'].unique()

array(['Unclear', 'Black', 'Asian', 'White', 'Latino', 'Native American'],
      dtype=object)

In [50]:
df['gender'].unique()

array(['F (identified as transgender, per police)', 'M', 'Male & Female',
       'F', 'Male', 'Female'], dtype=object)

In [51]:
df['gender']=df['gender'].apply(lambda x: 
                                x.replace('M','Male')
                                .replace('F','Female')
                                .replace('Female (identified as transgender, per police)', 'Transgender Female'))

df['gender'].unique()

array(['Transgender Female', 'Male', 'Maleale & Femaleemale', 'Female',
       'Maleale', 'Femaleemale'], dtype=object)

In [52]:
df['gender']=df['gender'].apply(lambda x: x.replace('Maleale & Femaleemale','Male & Female').replace('Maleale','Male').replace('Femaleemale','Female'))

df['gender'].unique()

array(['Transgender Female', 'Male', 'Male & Female', 'Female'],
      dtype=object)

We made a mess in this column but we have managed to clean our way out of it.

In [53]:
df['type'].unique()

array(['Mass', 'Spree'], dtype=object)

No issues here so let's carry on.

We are done cleaning our data. Since we do not want a very long notebook, we will save the cleaned data as a separate .csv file and analyze it in a separate notebook.

In [54]:
df.to_csv('showered_shootings.csv', index=False)

Now go freshen up and get ready to tackle the beast that brought us all here - data analysis!

## Thank you.